examples/webview2/src/main.rs

```diff
@@ -0,0 +1,622 @@
+        unsafe {
+            WindowsAndMessaging::SetWindowLongPtrA(
+                hwnd,
+                WINDOW_LONG_PTR_INDEX::GWLP_USERDATA,
+                Box::into_raw(webview) as _,
+            );
+        }
+    }
+
+    fn get_window_webview(hwnd: HWND) -> Option<Box<WebView>> {
+        unsafe {
+            let data =
+                WindowsAndMessaging::GetWindowLongPtrA(hwnd, WINDOW_LONG_PTR_INDEX::GWLP_USERDATA);
+
+            match data {
+                0 => None,
+                _ => {
+                    let webview_ptr = data as *mut WebView;
+                    let raw = Box::from_raw(webview_ptr);
+                    let webview = raw.clone();
+                    mem::forget(raw);
+
+                    Some(webview)
+                }
+            }
+        }
+    }
+}
+
+fn set_process_dpi_awareness() -> Result<()> {
+    let code = unsafe {
+        HiDpi::SetProcessDpiAwareness(PROCESS_DPI_AWARENESS::PROCESS_PER_MONITOR_DPI_AWARE)
+    };
+
+    if code.is_err() {
+        Err(Error::fast_error(code))
+    } else {
+        Ok(())
+    }
+}
+
+extern "system" fn window_proc(hwnd: HWND, msg: u32, w_param: WPARAM, l_param: LPARAM) -> LRESULT {
+    let webview = match WebView::get_window_webview(hwnd) {
+        Some(webview) => webview,
+        None => return unsafe { WindowsAndMessaging::DefWindowProcA(hwnd, msg, w_param, l_param) },
+    };
+
+    let frame = webview
+        .frame
+        .as_ref()
+        .expect("should only be called for owned windows");
+
+    match msg {
+        WindowsAndMessaging::WM_SIZE => {
+            let size = get_window_size(hwnd);
+            webview
+                .controller
+                .SetBounds(Rect {
+                    X: 0f32,
+                    Y: 0f32,
+                    Width: size.cx as f32,
+                    Height: size.cy as f32,
+                })
+                .expect("call SetBounds");
+            *frame.size.lock().expect("lock size") = size;
+            LRESULT(0)
+        }
+
+        WindowsAndMessaging::WM_CLOSE => {
+            unsafe {
+                WindowsAndMessaging::DestroyWindow(hwnd);
+            }
+            LRESULT(0)
+        }
+
+        WindowsAndMessaging::WM_DESTROY => {
+            webview.terminate().expect("window is gone");
+            LRESULT(0)
+        }
+
+        _ => unsafe { WindowsAndMessaging::DefWindowProcA(hwnd, msg, w_param, l_param) },
+    }
+}
+
+fn wait_with_pump<T>(rx: mpsc::Receiver<T>) -> Result<T> {
+    let mut msg = MSG::default();
+    let hwnd = HWND::default();
+
+    loop {
+        if let Ok(result) = rx.try_recv() {
+            return Ok(result);
+        }
+
+        unsafe {
+            match WindowsAndMessaging::GetMessageA(&mut msg, hwnd, 0, 0).0 {
+                -1 => {
+                    let error = format!("GetMessageW failed: {}", Win32::Debug::GetLastError());
+                    return Err(Error::new(ErrorCode::E_NOINTERFACE, &error));
```

Comment:
Is `E_NOINTERFACE` the right error?

Reply:
Probably not, but this was another example of trying to merge Error types.

Reply:
See https://github.com/microsoft/windows-rs/pull/647#discussion_r602316505